### PR TITLE
feat: graph document representation

### PR DESCRIPTION
**Pull request by davidbp**
`GraphDocument`  document type added with basic methods to store the graph inside the class.

Update: Added graph plot of the GraphDocument structure:

```
import jina
from jina.types.document.graph import GraphDocument
from jina import Document

g_doc = GraphDocument() 
node_0 = Document()     
node_1 = Document()
node_2 = Document()
g_doc.add_edge(node_0, node_2)
g_doc.add_edge(node_1, node_2)
g_doc.add_edge(node_0, node_1)
g_doc
```
Will show in a notebook:

Screenshot 2021-06-08 at 16 10 16

### CODE REVIEW

File: jina/proto/jina.proto [hunk 1]
```diff
@@ -67,6 +67,14 @@ message NamedScoreProto {
     string ref_id = 5; // the score is computed between doc `id` and `ref_id`
 }
 
+/**
+ * Represents the relevance model to `ref_id`
```
Comment (JoanFM) [inline jina/proto/jina.proto:71]: lets fix this message comment

File: jina/types/document/graph.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        document: Optional[DocumentSourceType] = None,
+        copy: bool = False,
+        **kwargs,
+    ):
+        super().__init__(document=document, copy=copy, **kwargs)
+        self._node_id_to_offset = {
+            node.id: offset for offset, node in enumerate(self.nodes)
+        }
+
+    def add_node(self, node: 'Document'):
```
Comment (JoanFM) [inline jina/types/document/graph.py:50]: lets put add methods together and remove methods together, and then properties together

File: jina/types/document/graph.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+        self._node_id_to_offset[node.id] = len(self.nodes)
+        self.nodes.append(node)
+
+    def remove_node(self, node: 'Document'):
+        """
+        Remove a node from the graph along with the edges that may contain it
+
+        :param node: the node to be removed from the graph
+        """
+        from scipy.sparse import coo_matrix
```
Comment (JoanFM) [inline jina/types/document/graph.py:69]: use ImportExtensions to see if scipy is installed and say that graph requires these dependencies

File: jina/types/document/graph.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+                    if col.item() == offset
+                ],
+                reference_doc=self,
+            )
+
+    def __iter__(self) -> Iterator[Tuple['Document']]:
+        for (row, col) in zip(self.adjacency.row, self.adjacency.col):
+            yield self.nodes[row.item()], self.nodes[col.item()]
+
+    def __len__(self) -> int:
```
Comment (JoanFM) [inline jina/types/document/graph.py:321]: to check if it makes sense

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/document/graph.py (hunk 3 above)
Comment (davidbp) [reply jina/types/document/graph.py:86]: Updated with ImportEntensions

File: jina/types/document/graph.py (hunk 4 above)
Comment (davidbp) [reply jina/types/document/graph.py]: Removed (and tests updated to respect this)

File: jina/types/document/graph.py [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+    def add_node(self, node: "Document"):
+        """
+        Add a a node to the graph
+
+        :param node: the node to be added to the graph
+        """
+        if node.id in self._node_id_to_offset:
+            default_logger.warning(f"Document {node.id} is already a node of the graph")
+            return
```
Comment (bwanglzu) [inline jina/types/document/graph.py:75]: not need. to return anything

File: jina/types/document/graph.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        """
+        Add an edge to the graph connecting `doc1` with `doc2`
+
+        :param doc1: the starting node for this edge
+        :param doc2: the ending node for this edge
+        :param features: Optional features dictionary to be added to this new created edge
+        """
+        from scipy.sparse import coo_matrix
+
+        for doc in [doc1, doc2]:
```
Comment (bwanglzu) [inline jina/types/document/graph.py:138]: why for loop here? isn't doc1 and doc2 a pair of document?

**Review by bwanglzu — COMMENTED** [review]
left some comments

File: jina/types/document/graph.py (hunk 5 above)
Comment (JoanFM) [reply jina/types/document/graph.py:75]: either this or put an `else`, but we do not want to add `Document` again

File: jina/types/document/graph.py (hunk 6 above)
Comment (JoanFM) [reply jina/types/document/graph.py:138]: yes, we could do `self.add_node` twice

Comment (davidbp) [reply jina/types/document/graph.py:138]: But there is still the need to loop right?

Comment (JoanFM) [reply jina/types/document/graph.py:138]: or one after the other

File: jina/types/document/graph.py [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+                ],
+                reference_doc=self,
+            )
+
+    def __iter__(self) -> Iterator[Tuple["Document"]]:
+        for (row, col) in zip(self.adjacency.row, self.adjacency.col):
+            yield self.nodes[row.item()], self.nodes[col.item()]
+
+    # def __len__(self) -> int:
+    #    return self.adjacency.getnnz()
```
Comment (bwanglzu) [inline jina/types/document/graph.py:339]: to be removed?

File: jina/types/document/graph.py [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+import numpy as np
+
+from . import Document, DocumentSourceType
+from ..arrays import ChunkArray
+from ..struct import StructView
+from ..ndarray.sparse.scipy import SparseNdArray
+from ...importer import ImportExtensions
+from ...logging.predefined import default_logger
+
+__all__ = ["GraphDocument"]
```
Comment (bwanglzu) [inline jina/types/document/graph.py:12]: minor thing
```suggestion
__all__ = ['GraphDocument']
```

File: jina/types/document/graph.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+    :param kwargs: further key value arguments
+    """
+
+    def __init__(
+        self,
+        document: Optional[DocumentSourceType] = None,
+        copy: bool = False,
+        **kwargs,
+    ):
+        GraphDocument._check_installed_array_packages()
```
Comment (bwanglzu) [inline jina/types/document/graph.py:46]:
```suggestion
        self._check_installed_array_packages()
```

File: jina/types/document/graph.py [hunk 10]
```diff
@@ -0,0 +1,339 @@
...
+                self._remove_edge_id(edge_id, f"{doc1.id}-{doc2.id}")
+
+    @property
+    def edge_features(self):
+        """
+        The dictionary of edge features, indexed by `edge_id` in the `edge list`
+
+        .. # noqa: DAR201
+        """
+        return StructView(self._pb_body.graph_info.edge_features)
```
Comment (bwanglzu) [inline jina/types/document/graph.py:204]: i suggest we rename `graph_info` to `graph` in protobuf:
```suggestion
        return StructView(self._pb_body.graph.edge_features)
```

File: jina/types/document/graph.py (hunk 6 above)
Comment (bwanglzu) [reply jina/types/document/graph.py:138]: okay

File: jina/types/document/graph.py (hunk 7 above)
Comment (JoanFM) [reply jina/types/document/graph.py:339]: yes

File: jina/types/document/graph.py (hunk 10 above)
Comment (JoanFM) [reply jina/types/document/graph.py:204]: As u prefer

File: jina/types/document/graph.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+
+if False:
+    from scipy.sparse import coo_matrix
+
+
+class GraphDocument(Document):
+    """
+    :class:`GraphDocument` is a data type created based on Jina primitive data type :class:`Document`.
+
+    It adds functionality that lets you work with a `Document` as a `graph` where all its chunks are the nodes in the `graph`.
```
Comment (florian-hoenicke) [inline jina/types/document/graph.py:22]: Maybe we could add to the documentation that it is a `directed graph`.

Comment (JoanFM) [reply jina/types/document/graph.py:22]: makes sense @DavidSanwald

Comment (davidbp) [reply jina/types/document/graph.py:22]: Fully Agree

File: jina/proto/jina.proto [hunk 12]
```diff
@@ -133,6 +141,8 @@ message DocumentProto {
     // List of evaluations performed on the document, each element corresponds to a metric
     repeated NamedScoreProto evaluations = 23;
 
+    // Related information to be used when a Document represents a graph with its chunks as nodes
+    Graph graph_info = 26;
```
Comment (hanxiao) [inline jina/proto/jina.proto:145]:
```suggestion
    GraphProto graph_info = 26;
```

Comment (davidbp) [reply jina/proto/jina.proto]: thanks, updated

File: jina/types/document/graph.py [hunk 13]
```diff
@@ -331,6 +331,41 @@ def get_incoming_nodes(self, doc: "Document") -> Optional[ChunkArray]:
                 reference_doc=self,
             )
 
+    @staticmethod
+    def load_from_dgl(dgl_graph: "DGLGraph"):
```
Comment (JoanFM) [inline jina/types/document/graph.py:335]: single quotes

File: jina/types/document/graph.py [hunk 14]
```diff
@@ -331,6 +331,41 @@ def get_incoming_nodes(self, doc: "Document") -> Optional[ChunkArray]:
...
+        Construct a GraphDocument from of graph with type `dgl.DGLGraph`
+
+        .. # noqa: DAR201
+        :param dgl_graph: the graph from which to construct a `GraphDocument`.
+        """
+        jina_graph = GraphDocument()
+        nodeid_to_doc = {}
+        for node in dgl_graph.nodes():
+            node_doc = Document()
+            nodeid_to_doc[int(node)] = node_doc
```
Comment (JoanFM) [inline jina/types/document/graph.py:346]: what is this `int(node)`. And what properties or features does this `Node` have?

File: jina/types/document/graph.py [hunk 15]
```diff
@@ -331,6 +331,41 @@ def get_incoming_nodes(self, doc: "Document") -> Optional[ChunkArray]:
...
+                nodeid_to_doc[int(node_source)], nodeid_to_doc[int(node_destination)]
+            )
+
+        return jina_graph
+
+    def to_dgl(self):
+        """
+        Construct a  `dgl.DGLGraph` from a `GraphDocument` instance.
+        """
+        import dgl
```
Comment (JoanFM) [inline jina/types/document/graph.py:360]: use `ImportExtensions`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/document/graph.py (hunk 14 above)
Comment (davidbp) [reply jina/types/document/graph.py:346]: it's a cast to standard python  integer from tensor tinteger.  The node has no properties or features.

File: jina/types/document/graph.py [hunk 16]
```diff
@@ -357,14 +357,38 @@ def to_dgl(self):
...
+
+                JINA_GLOBAL.torch_installed = True
+
         import torch
+        import dgl
 
         source_nodes = torch.tensor(self.adjacency.row.copy())
         destination_nodes = torch.tensor(self.adjacency.col.copy())
 
         return dgl.graph((source_nodes, destination_nodes))
```
Comment (JoanFM) [inline jina/types/document/graph.py:390]: I think we should add edge features frlm the graph proto and node dictionary as features to the nodes

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/document/graph.py (hunk 16 above)
Comment (JoanFM) [reply jina/types/document/graph.py:390]: no, lets keep it like this and in a next iteration we can see how it should work

**Review by JoanFM — APPROVED** [review]

File: extra-requirements.txt [hunk 17]
```diff
@@ -76,3 +76,4 @@ bs4:                        test
 aiostream:                  devel, cicd
 click:                      cicd
 jsonschema:                 cicd
+dgl==0.6.1:                 test, cicd
```
Comment (hanxiao) [inline extra-requirements.txt:79]:
```suggestion
dgl==0.6.1:                 cicd
```

File: jina/__init__.py [hunk 18]
```diff
@@ -100,6 +100,7 @@
 JINA_GLOBAL.scipy_installed = None
 JINA_GLOBAL.tensorflow_installed = None
 JINA_GLOBAL.torch_installed = None
+JINA_GLOBAL.dgl_installed = None
```
Comment (hanxiao) [inline jina/__init__.py:103]: seems unused

File: jina/types/document/graph.py [hunk 19]
```diff
@@ -0,0 +1,405 @@
...
+                required=False,
+                pkg_name='torch',
+                help_text=f'to_dgl_graph method requires torch to be installed',
+            ):
+                import torch
+
+                JINA_GLOBAL.torch_installed = True
+
+        import torch
+        import dgl
```
Comment (hanxiao) [inline jina/types/document/graph.py:396]: if importing them is a must, then `with ImportExtensions(required=True, ...)`

File: tests/unit/types/document/test_graph_document.py [hunk 20]
```diff
@@ -0,0 +1,179 @@
+import pytest
+
+from jina.types.document.graph import GraphDocument
```
Comment (hanxiao) [inline tests/unit/types/document/test_graph_document.py:3]: all double quote should be replaced to single quote in this file

**Review by hanxiao — CHANGES_REQUESTED** [review]
please change accordingly

**Review by JoanFM — APPROVED** [review]

**Pull request merged**